src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -1398,5 +1399,96 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
 
             return new ImmutableArray<T>(tmp);
         }
+
+        private struct MultiSet
+        {
+            private readonly IEqualityComparer<T> _equalityComparer;
+            private readonly Dictionary<int, LinkedList<(T value, int count)>> _dictionary = new();
```

Comment:
Since you're encoding a hashtable implementation here, wouldn't it make sense to use an array of buckets directly to avoid all the linked list allocations?

Reply:
Got it, fixed to array-based implementation. 
I referred the basic concept ([bucket headers array and buckets array](https://github.com/dotnet/runtime/blob/c1ee3cdbd123777384c0b9c4ffd704f691508605/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L40)) of generic `Dictionary` and `HashSet` and simplify something based on `ImmutableArray.RemoveRange` scenario, for example: no need to reuse free holes because `Remove()` will only happen after all `Add()` are called. 
Please correct my understanding.